### ISSUE TITLE
empty? bitset!, length?/empty? object! undefined

### ISSUE THREAD

**Issue by Kaj-de-Vos** [red/red#614]
All these are defined in R3:

red>> empty? charset []

**\* Script error: action 47 not defined for type: 28

red>> length? context []

**\* Script error: action 36 not defined for type: 30

red>> empty? context []

**\* Script error: action 47 not defined for type: 30

**Comment by hiiamboris** [comment]
It makes sense to define `empty? bitset` to be true if `false = find bitset n` for each `n` in the allowed range (is it 0 to 2^31-1 right now?).

**Comment by greggirwin** [comment]
I don't think a bitset is empty unless it contains no elements. 
```
>> make bitset! 0
== make bitset! #{}
>> make bitset! []
== make bitset! #{00}
```
Otherwise a block full of `false` values is also empty, right?

**Comment by hiiamboris** [comment]
Block is an array, bitset is a set of integers. A set is indeed empty if it has no integers in it, which can only reliably tested right now with `repeat i 2147483648 [if find bitset i - 1 [break/return 'not-empty]]`.

**Comment by hiiamboris** [comment]
Speaking of length, more formally correct definition of it would be a count of all set (i.e. 1) bits (also coherent with `empty?`). But I'm not sure where it will be practically useful yet.

**Comment by greggirwin** [comment]
A bitset is not a set of integers, it's a set of bits; individual binary digits / logic values.

As long as there is a value "slot" there, it's not empty, any more than a vector containing all zeros is. And there are other semantics that are different from series, e.g. `clear`, which sets all bits to 0 but does not deallocate the existing bits.

This opens a (worthwhile) conversation about the best fit of existing actions and funcs for bitsets. e.g. they support `length?`, which is is based on the implementation and doesn't mean the last `on` bit. They also support `pick`, but not `first/last` which are funcs that use `pick` internally. So I think this is a big picture design discussion.

**Comment by hiiamboris** [comment]
What you're describing is an implementation detail: how this set of integers is internally handled. I'm describing the model that this implementation aims to represent.

Current `clear` semantic seems suboptimal: there's no way to 'shorten' the bitset. But current `length?` semantic like I stated above carries no high level meaning. 

I don't think series (array) actions are a great fit for sets (because there's no head pointer), but we may indeed find some uses for them. An important aspect is to set/reset whole ranges is currently missing. Spec of `make bitset` has it, but only for newly created sets, not for the existing ones. Meaning e.g. to mark a range you first create a new bitset, then `or` it with an existing one, creating a third one in the process. Though it's only problematic with the current implementation, but not in principle, still it would be good to have a way to set/reset ranges. What action would fit that?

To block/hash/vector conversions make sense to me.

`insert`=`append` may work as union, but when given a block/hash/vector could read integers from it, just as it currently inserts strings into bitset (but chars in strings have a much more limited range). `remove/key` could be extended to block/hash/vector too.

`trim` could be used to "compact" the bitset (if its implementation doesn't provide automatic compaction).

Other actions doesn't seem to make sense at a quick glance.

**Comment by greggirwin** [comment]
> What you're describing is an implementation detail: how this set of integers is internally handled

No, I'm describing what it is, its general purpose, and how it's doc'd. What you think of as storing an integer, I think of as positional keys.

Being able to set ranges could be done via HOFs and slices, in a fancy world, but even now slowly and imperatively.

**Comment by hiiamboris** [comment]
Try to abstract from the implementation and see the `-set` part of `bitset` ;)
As for ranges, I realized it's another good use case for a range datatype, but until we have one a `pair!` will do, e.g. `insert bitset 100x200`.

**Comment by hiiamboris** [comment]
There could be another `bitstream!` or `bitarray!` datatype that would model what you're describing - an ordered sequence of individual bits - and could support all series actions. I don't think it'll be useful though. Compression and data exchange immediately come to mind, but who would want to implement them on (slow) Red level?

**Comment by hiiamboris** [comment]
I agree `first` and `last` will be useful if they return lowest and highest set bit number (just hit a use case).
